diff --git a/src/libcore/slice/mod.rs b/src/libcore/slice/mod.rs
index d89443662d..0e62dcd106 100644
--- a/src/libcore/slice/mod.rs
+++ b/src/libcore/slice/mod.rs
@@ -5025,8 +5025,11 @@ impl<A> SlicePartialEq<A> for [A]
         if self.len() != other.len() {
             return false;
         }
-        if self.as_ptr() == other.as_ptr() {
-            return true;
+        #[cfg(not(miri))] // FIXME Miri is not happy about this ptr comparison when empty slices are involved, so we skip it
+        {
+            if self.as_ptr() == other.as_ptr() {
+                return true;
+            }
         }
         unsafe {
             let size = mem::size_of_val(self);
