diff --git a/library/portable-simd/crates/core_simd/src/lib.rs b/library/portable-simd/crates/core_simd/src/lib.rs
index 64ba9705ef5..c1f19a6ca98 100644
--- a/library/portable-simd/crates/core_simd/src/lib.rs
+++ b/library/portable-simd/crates/core_simd/src/lib.rs
@@ -1,6 +1,7 @@
 #![no_std]
 #![feature(
     const_refs_to_cell,
+    const_intrinsic_copy,
     const_maybe_uninit_as_mut_ptr,
     const_mut_refs,
     convert_float_to_int,
@@ -11,7 +12,6 @@
     repr_simd,
     simd_ffi,
     staged_api,
-    stdsimd,
     strict_provenance,
     ptr_metadata
 )]
diff --git a/library/portable-simd/crates/test_helpers/src/lib.rs b/library/portable-simd/crates/test_helpers/src/lib.rs
index b80c745aaf2..eea9ba5baba 100644
--- a/library/portable-simd/crates/test_helpers/src/lib.rs
+++ b/library/portable-simd/crates/test_helpers/src/lib.rs
@@ -1,4 +1,4 @@
-#![feature(stdsimd, powerpc_target_feature)]
+#![feature(powerpc_target_feature)]
 
 pub mod array;
 

diff --git a/crates/std_detect/src/lib.rs b/crates/std_detect/src/lib.rs
index 19cc0217..d19128c2 100644
--- a/library/stdarch/crates/std_detect/src/lib.rs
+++ b/library/stdarch/crates/std_detect/src/lib.rs
@@ -23,7 +23,7 @@
 // Temporary hack: needed to build against toolchains from before the mass feature renaming.
 // Remove this as soon as the stdarch submodule is updated on nightly.
 #![allow(stable_features)]
-#![cfg_attr(not(feature = "rustc-dep-of-std"), feature(stdsimd))]
+#![cfg_attr(not(feature = "rustc-dep-of-std"), feature(stdarch_internal))]
 #![cfg_attr(
     all(
         any(target_arch = "x86", target_arch = "x86_64"),
