diff --git a/library/alloc/tests/arc.rs b/library/alloc/tests/arc.rs
index eb379e4d6a1..9191fa95150 100644
--- a/library/alloc/tests/arc.rs
+++ b/library/alloc/tests/arc.rs
@@ -212,6 +212,7 @@ fn hmm<'a>(val: &'a mut Weak<&'a str>) -> Weak<&'a str> {
 }
 
 #[test]
+#[cfg_attr(miri, ignore)] // test has UB (https://github.com/rust-lang/rust/issues/104565)
 fn arc_from_vec_opt() {
     let mut v = Vec::with_capacity(64);
     v.push(0usize);
diff --git a/library/alloc/tests/rc.rs b/library/alloc/tests/rc.rs
index 1d5f3c52006..0a19162a402 100644
--- a/library/alloc/tests/rc.rs
+++ b/library/alloc/tests/rc.rs
@@ -208,6 +208,7 @@ fn hmm<'a>(val: &'a mut Weak<&'a str>) -> Weak<&'a str> {
 }
 
 #[test]
+#[cfg_attr(miri, ignore)] // test has UB (https://github.com/rust-lang/rust/issues/104565)
 fn rc_from_vec_opt() {
     let mut v = Vec::with_capacity(64);
     v.push(0usize);
