diff --git a/src/liballoc/collections/vec_deque.rs b/src/liballoc/collections/vec_deque.rs
index d806bd0b1d..96f6ddab3d 100644
--- a/src/liballoc/collections/vec_deque.rs
+++ b/src/liballoc/collections/vec_deque.rs
@@ -2968,6 +2968,7 @@ mod tests {
     }
 
     #[test]
+    #[cfg(not(miri))] // FIXME: Miri fails, see https://github.com/rust-lang/rust/issues/60076.
     fn test_drain() {
         let mut tester: VecDeque<usize> = VecDeque::with_capacity(7);
 
diff --git a/src/liballoc/tests/slice.rs b/src/liballoc/tests/slice.rs
index ad2cd7c95e..23a18c7ee0 100644
--- a/src/liballoc/tests/slice.rs
+++ b/src/liballoc/tests/slice.rs
@@ -466,6 +466,7 @@ fn test_sort() {
 }
 
 #[test]
+#[cfg(not(miri))] // FIXME rand seems to perform invalid memory accesses.
 fn test_sort_stability() {
     #[cfg(not(miri))] // Miri is too slow
     let large_range = 500..510;
